secrets/vault.go

```diff
@@ -2,57 +2,33 @@ package secrets
 
 import (
 	"errors"
-	"io/ioutil"
-	"os"
-	"path"
 
 	"github.com/hashicorp/vault/api"
-	validator "gopkg.in/go-playground/validator.v9"
+	"github.com/hashicorp/vault/command/config"
 )
 
-var (
-	validate = validator.New()
-)
+func VaultFetchSecret(storepath, key string) (string, error) {
+	client, err := api.NewClient(nil)
+	if err != nil {
+		return "", err
+	}
 
-type vault struct {
-	token   string `validate:"required"`
-	address string `validate:"gt=1"`
-}
+	if client.Token() == "" {
+		helper, err := config.DefaultTokenHelper()
+		if err != nil {
+			return "", err
+		}
 
-// NewVaultFromEnv creates a vault SecretObtainer
-// using the usual environment settings that the vault cli would use.
-func NewVaultFromEnv() (*vault, error) {
-	v := &vault{
-		address: "http://localhost:8200",
-	}
-	if f, err := os.Open(path.Join(os.Getenv("HOME"), ".vault-token")); !os.IsNotExist(err) {
-		buff, err := ioutil.ReadAll(f)
+		token, err := helper.Get()
 		if err != nil {
-			return nil, err
+			return "", err
 		}
-		v.token = string(buff)
-	}
-	if token, set := os.LookupEnv("VAULT_TOKEN"); set && len(v.token) == 0 {
-		v.token = token
-	}
-	if addr, set := os.LookupEnv("VAULT_ADDR"); set {
-		v.address = addr
-	}
-	return v, validate.Struct(v)
-}
 
-func VaultFetchSecret(storepath, key string) (string, error) {
-	v, err := NewVaultFromEnv()
-	if err != nil {
-		return "", err
-	}
-	client, err := api.NewClient(&api.Config{
-		Address: v.address,
-	})
-	if err != nil {
-		return "", err
+		if token != "" {
```

Comment:
I don't think this `if` is necessary, but it's consistent with how vault do it here: https://github.com/hashicorp/vault/blob/df18871704fe869e9be45b542a6b1eb2fe46c293/command/base.go#L98

So I erred on keeping to their style - perhaps it's a bit more explicit?

Reply:
I think this is a worthwhile change 👍 

I need to confirm that it will load a vault token store in `~/.vault-token` but I really do appreciate making this simpler.

What I'd like to do in future is change the token obtain method to support token leasing from vault given a kubernetes service account and role.